backfill_corrections/delphiBackfillCorrection/R/model.R

```diff
@@ -262,3 +267,63 @@ generate_filename <- function(indicator, signal,
   )
   return(filename)
 }
+
+#' Get date range of data to use for training models
+#'
+#' Calculate training start and end dates based on user settings.
+#' `training_start_date` is the minimum allowed target date when selecting
+#' training data to use. `training_end_date` is the maximum allowed target
+#' date and maximum allowed issue date.
+#'
+#' Cases:
+#'   1. We are training new models.
+#'   2. We are not training new models and cached models exist.
+#'   3. We are not training new models and cached models don't exist.
+#'
+#' Sometimes we want to allow the user to specify an end date in
+#' params that overrides the automatically-generated end date. This is
+#' only relevant when the user requests to train new models.
+#'
+#' @template params-template
+get_training_date_range <- function(params) {
+  if (params$train_models) {
+    if (params_element_exists_and_valid(params, "training_end_date")) {
+      # Use user-provided end date.
+      training_end_date <- as.Date(params$training_end_date)
+    } else {
+      # Default end date is today.
+      training_end_date <- TODAY
+    }
+  } else {
+    # Get end date from cached model files. Assumes filename format like
+    # `20220628_20220529_changehc_covid_state_lambda0.1_count_ca_lag5_tau0.9.model`
+    # where the leading date is the training end date for that model, and the
+    # second date is the training start date.
+    model_files <- list.files(params$cache_dir, "202[0-9]{5}_202[0-9]{5}*.model")
```

Comment:
I always run into problem with this line. R told me that 

```
Error in list.files(params$cache_dir, "202[0-9]-[0-9]{2}-[0-9]{2}*.model") : 
  invalid 'pattern' regular expression
```

Reply:
Sorry, I have a fix for that but haven't pushed yet.